File: hard/52. N-Queens II/52. N-Queens II.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        # Clues:
        # 1. cols set with queens
        # 2. positive and negative diagonals with queens
        # The rule for positive diagonal: row + column
        # The rule for negative diagonal: row - column

        col = set()
        posDiag = set()
        negDiag = set()

        self.res = 0

        def backtrack(r: int) -> None:
            if r == n:
                self.res += 1
                return

            for c in range(n):
                if c in col or (r + c) in posDiag or (r - c) in negDiag:
                    continue

                col.add(c)
                posDiag.add(r + c)
                negDiag.add(r - c)

                backtrack(r + 1)

                col.remove(c)
                posDiag.remove(r + c)
                negDiag.remove(r - c)

        backtrack(0)
        return self.res

File: hard/52. N-Queens II/52. N-Queens II.ts
function totalNQueens(n: number): number {
    const col = new Set()
    const positiveDiag = new Set() // (r + c)
    const negativeDiag = new Set() // (r - c)

    let res: number = 0

    const backtrack = (r: number): void => {
        if (r === n) {
            res += 1
            return
        }

        for (let c = 0; c < n; c++) {
            if (col.has(c) || positiveDiag.has(r + c) || negativeDiag.has(r - c))
                continue

            col.add(c)
            positiveDiag.add(r + c)
            negativeDiag.add(r - c)

            backtrack(r + 1)

            col.delete(c)
            positiveDiag.delete(r + c)
            negativeDiag.delete(r - c)
        }
    }

    backtrack(0)
    return res
};